grant-project/grant-monorepo
# Backend Proposal Reads Pt. 1

Closes #124.

#### What this does
- integrates contract data into GET `proposals/<id>` and GET `proposals`
- refactors FE to use the new data
- `crowdFundContract` has been removed from the redux store, we should do the same for any other non-serializables (web3 comes to mind)
- `lib/crowdFundContracts.ts` lazy-loads and caches contracts now, called from anything that wants a contract and knows its address & has a web3 ref available
- browse proposals and proposal details no longer require web3
- `MetaMaskRequiredButton` is used in the campaign block to guide the user to use MM, unlock & set proper network (based on `Template` with web3 required) if they don't have MM

### Testing
- `pip install -r requirements/dev.txt`
- get your dev env running
- **with MetaMask**
  - create several proposals
  - attempt funding, voting, cancelation (this will be easier with the e2e update)
- **without MetaMask**
  - make sure you can browse proposals & details
  - make sure the campaign block has the "MetatMask required" button
  - enable MM and set to wrong network, ensure the campaign block has proper message
  - set MM to corrent network, leave it locked, ensure locked message is displayed
  - unlock and click link in campaign block message, ensure funding becomes enabled

##### No MetaMask
![screen shot 2018-11-14 at 10 42 35 am](https://user-images.githubusercontent.com/11430954/48500904-8f349780-e801-11e8-97f3-085a5305edf2.png)

##### Account Locked
![screen shot 2018-11-14 at 10 43 01 am](https://user-images.githubusercontent.com/11430954/48500936-a5daee80-e801-11e8-8ee3-1c2e10ce18b1.png)

##### Wrong Network
![screen shot 2018-11-14 at 10 43 36 am](https://user-images.githubusercontent.com/11430954/48500955-af645680-e801-11e8-9e70-6db5534c6aa0.png)
